### PR TITLE
Video + Audio datatypes

### PR DESCRIPTION
**Pull request by hexylena**
Opening this to coordinate work amongst the various groups working on video (/audio?) plugins.

Pinging @gregvonkuster, @stevecassidy @yvanlebras, would love for you all to contribute to this. I think I've got a good base "Video" class, uses ffprobe for detecting file types. Thus is can detect things like number of streams, resolution, framerate, and detailed information about every stream (should we wish to record more metadata)

E.g. we could easily include metadata like https://github.com/Alveo/docker-galaxy-alveo/blob/master/datatypes/speech.py#L20 since ffprobe + my datatype will properly recognise those files as containing some audio streams and no video streams.

### CODE REVIEW

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]: The try/except is handled by the framework, no need to do this again here. (and all the other places in this PR)

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]: This is deprecated, no need to do that anymore.

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]: If there is a context manager available we should use it, otherwise this should be wrapped in try/finally

File: lib/galaxy/datatypes/speech.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/speech.py]:
```suggestion
            text = fd.read(len(self.header))
```

File: lib/galaxy/datatypes/speech.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/speech.py]: This isn't tight enough to my liking, would you be OK with dropping the sniffing for this datatype (if there's no way to make it more tight) ?

File: config/datatypes_conf.xml.sample
Comment (mvdbeek) [inline config/datatypes_conf.xml.sample]: I think the par datatype is missing here.

File: lib/galaxy/datatypes/speech.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/speech.py]: This and the logging import are unused AFAICT (also in media.py)

File: lib/galaxy/datatypes/media.py
Comment (hexylena) [reply lib/galaxy/datatypes/media.py]: Sounds good!

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py:167]: This means it'll not fail setting metadata even if wave.Error is raised, is that intentional ? (which is fine, the metadata elements are marked as optional, just checking since it wasn't in the previous code.)

Comment (hexylena) [reply lib/galaxy/datatypes/media.py:167]: Yeah, if it's been successfully sniffed but metadata setting fails for some reason? That seems reasonable?

Comment (mvdbeek) [reply lib/galaxy/datatypes/media.py:167]: Sure, I guess it'll have an impact on how you write tools. If it's optional the tool author should probably check if the attributes are set if there is some logic using these in the tool. Either way is fine with me.

File: lib/galaxy/datatypes/speech.py
Comment (hexylena) [reply lib/galaxy/datatypes/speech.py]: I've restricted it in a new commit.

File: lib/galaxy/datatypes/speech.py
Comment (nsoranzo) [inline lib/galaxy/datatypes/speech.py]: This will return after the checking only the first line, not sure this is intentional (since you are using a for loop). There are 9 compulsory headers, we can check that all are present.

File: config/datatypes_conf.xml.sample
Comment (hexylena) [reply config/datatypes_conf.xml.sample]: fixed

**Comment by mvdbeek** [comment]
I'm expecting the integration tests to fail due to lack of `ffprobe`, I'll update the image.

File: lib/galaxy/datatypes/speech.py
Comment (hexylena) [reply lib/galaxy/datatypes/speech.py]: Indeed. in the test case files submitted by @stevecassidy the mandatory headers were missing. Not sure how best to handle that.

**Comment by hexylena** [comment]
Thanks @mvdbeek

File: config/datatypes_conf.xml.sample
Comment (mvdbeek) [inline config/datatypes_conf.xml.sample]:
```suggestion
    <datatype extension="TextGrid" type="galaxy.datatypes.speech:TextGrid" display_in_upload="true" mimetype="text/plain"/>
```
(or `textgrid` here and change file_ext in the class)

Comment (nsoranzo) [reply config/datatypes_conf.xml.sample]: Please revert and change the `file_ext` in the class, all extension should be lowercase see e.g. https://github.com/galaxyproject/galaxy/commit/376212eff900d28450165ddf715f3d9a29870ca9

File: lib/galaxy/datatypes/speech.py
Comment (mvdbeek) [reply lib/galaxy/datatypes/speech.py]: Since it is the sniffer I'd err on the side of caution ... if you really have a file that is missing mandatory lines you can always assign the datatype. But the `set_metadata` code has the same issue, so I'm not sure what the intention was there.

File: config/datatypes_conf.xml.sample
Comment (hexylena) [reply config/datatypes_conf.xml.sample]: Can do, I was not aware of that restriction. Thanks @nsoranzo

**Comment by hexylena** [comment]
Tagged WIP as awaiting the visualisation addition from @anuprulez.

File: lib/galaxy/datatypes/speech.py
Comment (nsoranzo) [inline lib/galaxy/datatypes/speech.py]:
```suggestion
    file_ext = "textgrid"
```

Comment (hexylena) [reply lib/galaxy/datatypes/speech.py]: Huh, made change and forgot to stage. This works, thanks.

File: lib/galaxy/datatypes/speech.py
Comment (nsoranzo) [inline lib/galaxy/datatypes/speech.py]: Remove this function? We run unit tests in Travis.

File: lib/galaxy/datatypes/speech.py
Comment (hexylena) [reply lib/galaxy/datatypes/speech.py]: Sounds good. We can assert those lines are present in first 9 lines.

Comment (nsoranzo) [reply lib/galaxy/datatypes/speech.py]: Not sure if they must be in the first 9 lines since there are other optional headers.

Comment (hexylena) [reply lib/galaxy/datatypes/speech.py]: Yeah, I re-read the spec and implemented to check first 30 lines due to 21 optional headers.

Comment (hexylena) [reply lib/galaxy/datatypes/speech.py]: The implementation is bad, reworking it now.

**Comment by hexylena** [comment]
Test failures are relevant, I'm working on them.

**Comment by mvdbeek** [comment]
https://github.com/erasche/galaxy/pull/22 should allow for the integration tests to pass. I've just added ffmpeg like this:
```
FROM galaxy/testing-base:19.01.0

RUN apt-get update && apt-get install --no-install-recommends -y ffmpeg
```

The right thing would be to rebuild from scratch, which I will do in a separate PR.

**Comment by mvdbeek** [comment]
`test/integration/test_datatype_upload.py::test_upload_datatype_auto[video_1.mp4] PASSED [ 26%]`

**Comment by mvdbeek** [comment]
I've added the WIP tag, we shouldn't merge this with the temporary image in place.

**Comment by mvdbeek** [comment]
The travis tests are failing since there's no ffprobe installed. That said I don't think the sniffing tests are absolutely necessary, the integration tests are a lot more thorough.

**Comment by hexylena** [comment]
I wish I had a more pythonic way to do this, but ffprobe gave the information that I needed (streams available, framerate), system dependencies are quite an unpleasant burden.

**Comment by mvdbeek** [comment]
ffmpeg is also not a particularly small package, but I think it's OK.

**Comment by nsoranzo** [comment]
May something like [ffmpeg-python](https://pypi.org/project/ffmpeg-python/) be helpful?

**Comment by hexylena** [comment]
@nsoranzo I stumbled across that earlier today, but unfortunately it does [the same thing as me](https://github.com/kkroening/ffmpeg-python/blob/d47890aebd51add447b410f7e9e7dd6baac1e346/ffmpeg/_probe.py)

**Comment by nsoranzo** [comment]
@erasche I just hoped it would simplify installation, but it seems to assume ffmpeg is already installed.

**Comment by nsoranzo** [comment]
The lighter alternative could be to give up detailed audio/video metadata and use [filetype](https://pypi.org/project/filetype/) to sniff the formats.

**Comment by hexylena** [comment]
The metadata was intended to allow proper video playback. I identified that I could not step frame by frame through videos without the `fps` at minimum.

The new video plugin does not support this feature yet, and the old one only did with a videojs plugin I wrote, but I'm somewhat reluctant to downgrade the plugin and remove potentially useful metadata?

**Comment by mvdbeek** [comment]
But we could still use filetype for sniffing. That should have a slight speed advantage and works everywhere.
I suppose at some point visualizations (not implemented yet) and tools could then declare a dependency on ffmpeg if they require more detailed metadata ?

**Comment by hexylena** [comment]
> But we could still use filetype for sniffing. That should have a slight speed advantage and works everywhere.

true, that's probably a good idea.

> I suppose at some point visualizations (not implemented yet) and tools could then declare a dependency on ffmpeg if they require more detailed metadata ?

Here "(not implemented yet)" referred to the proposed functionality that tools could declare dependencies if they require more metadata? Just so I don't misunderstand

**Comment by nsoranzo** [comment]
> But we could still use filetype for sniffing. That should have a slight speed advantage and works everywhere.

Indeed.
@erasche You need to add try/except to handle the case where ffmpeg is not installed any way.

**Comment by hexylena** [comment]
@nsoranzo 

> You need to add try/except to handle the case where ffmpeg is not installed any way

Per https://github.com/galaxyproject/galaxy/pull/4949#discussion_r261141329 I removed most try/excepts.

**Comment by mvdbeek** [comment]
Well, you don't need try/except for sniffing, just for set_metadata (assuming files without the metadata are usable)

**Comment by hexylena** [comment]
Ok, thanks for suggestions @mvdbeek @nsoranzo, I'll get that patched.

File: lib/galaxy/datatypes/media.py
Comment (nsoranzo) [reply lib/galaxy/datatypes/media.py]: > The try/except is handled by the framework, no need to do this again here. (and all the other places in this PR)

For future reference, that's true only for `sniff()` methods, not `set_metadata()`.

**Comment by mvdbeek** [comment]
@erasche the new testing image with ffmpeg is up, if you rebase / merge with dev you should get the right image (galaxy/testing-base:19.05.0).

**Comment by annefou** [comment]
So cool to have video (/audio) formats in Galaxy! We (Climate community) are very interested; this is very useful to show the evolution of a physical phenomena (for instance to show on a map the evolution of the temperature over the last century, to show changes in vegetatation, etc.).
We are using a range of formats (for instance mp4, mov, avi).

**Comment by bgruening** [comment]
@mvdbeek I see some test failing with ffprobe missing, e.g. test_galaxy_packages (https://circleci.com/gh/galaxyproject/galaxy/72213?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link). Do we need to include ffprobe somewhere else?

**Comment by mvdbeek** [comment]
No, I don't think we want additional hard dependencies. Use of `ffprobe` should be optional. Those tests should be skipped if no ffprobe is on path.

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
    data = json.loads(unicodify(subprocess.check_output(['ffprobe', '-loglevel', 'quiet', '-show_format', '-show_streams', '-of', 'json', path])))
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py:9]:
```suggestion
from galaxy.util import unicodify

```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
        with open(filename, 'rb') as fh:
            wave.open(fh)
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
```

File: .circleci/config.yml
Comment (mvdbeek) [inline .circleci/config.yml]: What you get when you do `pip install wave` is https://pypi.org/project/Wave/#history, which I don't think is what you're aiming for?
You're importing `wave`, which comes from the standard library (https://docs.python.org/3/library/wave.html), so I think you can remove the wave installation and conditional import

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
import wave
```

File: config/plugins/visualizations/video_player/static/js/plyr.js
Comment (mvdbeek) [inline config/plugins/visualizations/video_player/static/js/plyr.js]: I am not sure if this is still the way we add javascript files.
I think we usually add a package.json file with the dependencies and the build command, but on this front I think we need help from @dannon or @guerler.

File: .github/workflows/integration.yaml
Comment (mvdbeek) [inline .github/workflows/integration.yaml]:
```suggestion
```

File: .github/workflows/integration.yaml
Comment (mvdbeek) [inline .github/workflows/integration.yaml:59]:
```suggestion
      run: sudo apt-get update && sudo apt-get install ffmpeg -y
      if: matrix.subset == 'upload_datatype'
```

File: .circleci/config.yml
Comment (mvdbeek) [inline .circleci/config.yml]:
```suggestion
```

File: .circleci/config.yml
Comment (mvdbeek) [inline .circleci/config.yml]:
```suggestion
```

File: .circleci/config.yml
Comment (mvdbeek) [inline .circleci/config.yml]:
```suggestion
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
        with wave.open(filename, 'rb') as fp:
            return True
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
        try:
            with wave.open(dataset.dataset.file_name, 'rb') as fd:
                dataset.metadata.rate = fd.getframerate()
                dataset.metadata.nframes = fd.getnframes()
                dataset.metadata.sampwidth = fd.getsampwidth()
                dataset.metadata.nchannels = fd.getnchannels()
        except wave.Error:
            pass
```

File: lib/galaxy/datatypes/speech.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/speech.py]:
```suggestion
            text = fd.read(len(self.header))
            return text == self.header
```

File: lib/galaxy/datatypes/speech.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/speech.py]:
```suggestion
```

File: lib/galaxy/datatypes/speech.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/speech.py]:
```suggestion
```

File: lib/galaxy/datatypes/speech.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/speech.py]:
```suggestion
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
```

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
```

**Review by mvdbeek — COMMENTED** [review]
@bgruening @hexylena I guess we missed that because the PR was still labelled WIP.
I think we can merge this if you can apply my suggestions.

**Comment by mvdbeek** [comment]
Ah, and the mp3 and mp4 test files should be smaller.

File: lib/galaxy/datatypes/media.py
Comment (hexylena) [reply lib/galaxy/datatypes/media.py]: jeez, why did I do this so much

**Comment by hexylena** [comment]
@mvdbeek sure I can fix those files.

File: config/plugins/visualizations/video_player/static/js/plyr.js
Comment (mvdbeek) [reply config/plugins/visualizations/video_player/static/js/plyr.js]: Ah, up to date instructions are at https://training.galaxyproject.org/training-material/topics/dev/tutorials/visualization-charts/tutorial.html.

**Comment by hexylena** [comment]
I've rebased/force pushed to fix the too-large test cases. @anuprulez would you be able to fix the visualisation per @mvdbeek's comment?

File: lib/galaxy/datatypes/media.py
Comment (mvdbeek) [inline lib/galaxy/datatypes/media.py]:
```suggestion
        with wave.open(filename, 'rb'):
```

**Comment by dannon** [comment]
I took care of the packaging fixes for the viz plugin.

**Comment by mvdbeek** [comment]
Awesome, I'll take care of the rest.

**Comment by mvdbeek** [comment]
@dannon we can drop config/plugins/visualizations/audio_player/static/css/plyr.css and and config/plugins/visualizations/audio_player/static/js/plyr.js as well, right ?

**Comment by dannon** [comment]
@mvdbeek Whoops; I only looked at the video player.  I'll follow up.

**Comment by mvdbeek** [comment]
Yeah, I think a merged media player is fine.

**Comment by dannon** [comment]
I merged the sources and build process for a generic `media_player` but kept the Galaxy-side plugins separate; should be good to go now.

**Review by mvdbeek — APPROVED** [review]
Alright, this looks great. Unfortunately I'll have to squash the merge commit so that we can keep the large test data out of the commit history.

**Comment by dannon** [comment]
@mvdbeek Instead of that, how about we rebase to prune it out?   I can do it.

**Comment by bgruening** [comment]
Thanks all for working on this!!!

**Comment by bgruening** [comment]
Btw. I'm ok if you remove my commits.

**Comment by hexylena** [comment]
Awesome! Thanks all.

**Pull request merged**
